src/Dhall/Core.hs

```diff
@@ -1375,75 +1403,70 @@ normalize e = case e of
-        kts' = fmap normalize kts
+        kts' = fmap (loop) kts
     UnionLit k v kvs -> UnionLit k v' kvs'
       where
-        v'   =      normalize v
-        kvs' = fmap normalize kvs
+        v'   =      loop v
+        kvs' = fmap (loop) kvs
     Combine x0 y0 ->
         let combine x y = case x of
                 RecordLit kvsX -> case y of
                     RecordLit kvsY ->
                         let kvs = Data.Map.unionWith combine kvsX kvsY
-                        in  RecordLit (fmap normalize kvs)
+                        in  RecordLit (fmap (loop) kvs)
                     _ -> Combine x y
                 _ -> Combine x y
-        in  combine (normalize x0) (normalize y0)
+        in  combine (loop x0) (loop y0)
     Prefer x y ->
         case x' of
             RecordLit kvsX ->
                 case y' of
                     RecordLit kvsY ->
-                        RecordLit (fmap normalize (Data.Map.union kvsY kvsX))
+                        RecordLit (fmap (loop) (Data.Map.union kvsY kvsX))
                     _ -> Prefer x' y'
             _ -> Prefer x' y'
       where
-        x' = normalize x
-        y' = normalize y
+        x' = loop x
+        y' = loop y
     Merge x y t      ->
         case x' of
             RecordLit kvsX ->
                 case y' of
                     UnionLit kY vY _ ->
                         case Data.Map.lookup kY kvsX of
-                            Just vX -> normalize (App vX vY)
+                            Just vX -> loop (App vX vY)
                             Nothing -> Merge x' y' t'
                     _ -> Merge x' y' t'
             _ -> Merge x' y' t'
       where
-        x' =      normalize x
-        y' =      normalize y
-        t' = fmap normalize t
+        x' =      loop x
+        y' =      loop y
+        t' = fmap (loop) t
     Field r x        ->
-        case normalize r of
+        case loop r of
             RecordLit kvs ->
                 case Data.Map.lookup x kvs of
-                    Just v  -> normalize v
-                    Nothing -> Field (RecordLit (fmap normalize kvs)) x
+                    Just v  -> loop v
+                    Nothing -> Field (RecordLit (fmap (loop) kvs)) x
             r' -> Field r' x
-    Note _ e' -> normalize e'
+    Note _ e' -> loop e'
     Embed a -> Embed a
-  where
-    -- This is to avoid a `Show` constraint on the @a@ and @s@ in the type of
-    -- `normalize`.  In theory, this might change a failing repro case into
-    -- a successful one, but the risk of that is low enough to not warrant
-    -- the `Show` constraint.  I care more about proving at the type level
-    -- that the @a@ and @s@ type parameters are never used
-    e'' = bimap (\_ -> ()) (\_ -> ()) e
 
-    text = "normalize (" <> Data.Text.pack (show e'') <> ")"
+-- | Use this to wrap you embedded functions (see `normalizeWith`) to make them
+--   polymorphic enough to be used.
+type CtxFun a = forall s. Expr s a -> Expr s a
```

Comment:
Perhaps inline the `CtxFun` type synonym into the type of `normalizeWith` since it's only used once